src/codelabs/null-safety.md

```diff
@@ -0,0 +1,362 @@
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: More nullable types!
+
+Type parameters for generics can also be nullable or non-nullable. Try using
+question marks to correct the type declarations of `aNullableListOfStrings` and
+`aListOfNullableStrings`:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/more_nullable_types.dart" replace="/String\?/String/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  List<String> aListofStrings = ['one', 'two', 'three'];
+  List<String> aNullableListOfStrings = [];
+  List<String> aListofNullableStrings = ['one', null, 'three'];
+
+  print('aListofStrings is $aListofStrings.');
+  print('aNullableListOfStrings is $aNullableListOfStrings.');
+  print('aListofNullableStrings is $aListofNullableStrings.');
+}
+```
+
+### Example: Definite assignment
+
+Dart's type system is crafty enough to track where variables are assigned and
+where their values are read, and to verify that non-nullable fields are given
+values before any code tries to read from them. This process is called
+"definite assignment."
+
+Try uncommenting the if-else statement in the code below, and
+watch the analyzer errors disappear:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  String text;
+
+  //if (DateTime.now().hour < 12) {
+  //  text = "It's morning! Let's make aloo paratha!";
+  //} else {
+  //  text = "It's afternoon! Let's make biryani!";
+  //}
+
+  print(text);
+  print(text.length);
+}
+```
+
+## New null-aware operators: ? and !
+
+Null safety introduces two operators, ? and !. Here's how you use them:
+
+-   If you're creating a variable that can have the value null, add a
+    question mark (?) to the variable's type declaration. Types are
+    non-nullable by default and made nullable by adding ?.
+-   If you're accessing a property and the operand might be null, use the
+    null-aware version member access operator (?.).
+-   If you're sure that an expression with a nullable type isn't null, you can 
+    add a postfix exclamation mark (!) to make Dart treat it as non-nullable.
+
+### Example: Conditional access
+
+Conditional access is a handy way to tighten up code that needs to read
+properties that could be null. For example, the following expression evaluates
+to the value of `b` as long as `a` isn't null. If `a` is null, then the
+expression evaluates to null:
+
+  a?.b
+
+Without null safety, *two* conditional access operators are needed for the
+following code to work: one after `big` and one after `little`. With null
+safety, conditional access can short-circuit, so this expression only requires a
+single `?`:
```

Comment:
This example is a little hard to understand because it's so hypothetical. We've already added `?` to a type up above, and we don't show `?.`, so it's unclear what new thing this is meant to show.

Reply:
yeah i think it's best to remove the conditional access section.

Reply:
yeah i agree. i think i'll remove this section.